What's new in .NET 11 – Preview 4 docs update (#53689)
dotnet/docs: docs/core/whats-new/dotnet-11/runtime.md

> Starting with Preview 3, a `net11.0` project no longer requires `<EnablePreviewFeatures>true</EnablePreviewFeatures>` to use Runtime Async.

Reviewer: ```suggestion
A `net11.0` project no longer requires `<EnablePreviewFeatures>true</EnablePreviewFeatures>` to use Runtime Async.
```

> Starting with Preview 4, the .NET runtime libraries themselves are compiled with `runtime-async=on`. The runtime libraries no longer contain compiler-generated state machines and rely entirely on runtime-provided async. This makes it possible to migrate an entire app (with only library dependencies) to the new model, and it provides broad functional and performance validation of the feature. The product team welcomes any reports—positive or negative—about throughput and library size changes you observe.

Reviewer: ```suggestion
The .NET runtime libraries themselves are compiled with `runtime-async=on`. The runtime libraries no longer contain compiler-generated state machines and rely entirely on runtime-provided async. This makes it possible to migrate an entire app (with only library dependencies) to the new model, and it provides broad functional and performance validation of the feature. The product team welcomes any reports—positive or negative—about throughput and library size changes you observe.
```

> Preview 4 also includes two additional improvements:

Reviewer: ```suggestion
.NET 11 also includes two additional improvements:
```

- **Devirtualization in ReadyToRun images:** ReadyToRun (R2R) images can now devirtualize non-shared generic virtual method calls, improving performance of ahead-of-time compiled code for generic scenarios.
- **SVE2 intrinsics:** New Arm SVE2 (Scalable Vector Extension 2) intrinsics are available: `ShiftRightLogicalNarrowingSaturate(Even|Odd)`. These expand the set of vectorized operations available on Arm hardware that supports SVE2.

> For better performance and code quality, Preview 4 adds several more JIT optimizations:

Reviewer: ```suggestion
For better performance and code quality, .NET 11 adds several more JIT optimizations:
```

static bool IsAdmin(string role) => role == "Admin";
```

The optimization applies to string literals, `const string` fields, and UTF-8 literals (for example, `"PNG"u8`).

### Eliminating bounds checks after an empty-span guard

The JIT now picks up the `length != 0` assertion from an empty-span check and uses it to prove the bounds check on the first element succeeds:

```csharp
if (!span.IsEmpty && span[0] == value)
{
    // The bounds check on span[0] is now eliminated.
}
```

### Redundant branch and test elimination

When an outer predicate is already implied by an inner branch, the JIT now removes the redundant outer check. Similarly, when code assigns a value in a conditional and then immediately tests it, the second test is eliminated:

```csharp
if (x > 0)
{
    if (x > 1) S();  // The outer x > 0 check is folded away.
}

int y = condition ? 1 : 2;
if (y == 1) A(); else B();  // The y == 1 test is eliminated;
                             // each branch of the ternary goes directly to A() or B().
```

These optimizations are most visible after inlining, where guards from different methods land in the same compiled body.

## Hardware intrinsics and code generation

> Preview 4 includes several new hardware intrinsics and code generation improvements:

Reviewer: ```suggestion
.NET 11 includes several new hardware intrinsics and code generation improvements:
```

- **Better debugging symbols:** Symbol and stack trace quality for WebAssembly debugging has improved, making it easier to diagnose issues in browser-hosted .NET apps.
- **`float[]`, `Span<float>`, and `ArraySegment<float>` marshaling:** `float[]`, `Span<float>`, and `ArraySegment<float>` are now marshaled more directly across JavaScript boundaries, reducing overhead for interop-heavy code.

> Preview 4 adds significant CoreCLR-on-WebAssembly progress:

Reviewer: ```suggestion
.NET includes improvements to CoreCLR-on-WebAssembly:
```
